plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs

```diff
@@ -0,0 +1,495 @@
+  , SomeTypeIn DefaultUniUnit
+  , SomeTypeIn DefaultUniBool
+  , SomeTypeIn DefaultUniByteString
+  , SomeTypeIn DefaultUniString
+  -- TODO: should we have more examples of lists and pairs here?
+  , SomeTypeIn $ DefaultUniList DefaultUniBool
+  , SomeTypeIn $ DefaultUniPair DefaultUniInteger DefaultUniUnit]
+builtinTysAt (Star :-> Star) =
+  [ SomeTypeIn DefaultUniProtoList
+  , SomeTypeIn $ DefaultUniProtoPair `DefaultUniApply` DefaultUniString
+  ]
+builtinTysAt (Star :-> Star :-> Star) = [SomeTypeIn DefaultUniProtoPair]
+builtinTysAt _ = []
+
+-- | Generate "small" types at a given kind such as builtins, bound variables, bound datatypes,
+-- and lambda abstractions \ t0 ... tn. T
+genAtomicType :: Kind () -> GenTm (Type TyName DefaultUni ())
+genAtomicType k = do
+  tys <- asks geTypes
+  dts <- asks geDatas
+  let atoms = [ TyVar () x | (x, k') <- Map.toList tys, k == k' ] ++
+              [ TyVar () x | (x, Datatype _ (TyVarDecl _ _ k') _ _ _) <- Map.toList dts, k == k' ]
+      builtins = map (TyBuiltin ()) $ builtinTysAt k
+      lam k1 k2 = do
+        x <- genMaybeFreshTyName "a"
+        TyLam () x k1 <$> bindTyName x k1 (genAtomicType k2)
+  -- TODO: probably should be 'frequencyTm'?
+  oneof $ concat
+    [ map pure atoms
+    , [elements builtins | not $ null builtins]
+    , [lam k1 k2 | KindArrow _ k1 k2 <- [k]]
+    ]
+
+-- | Generate a type at a given kind
+genType :: Kind () -> GenTm (Type TyName DefaultUni ())
+genType k = checkInvariants $ onSize (min 10) $
+  ifSizeZero (genAtomicType k) $
+    frequency $
+      [ (5, genAtomicType k) ] ++
+      [ (10, genFun) | k == Star ] ++
+      [ (5, genForall) | k == Star ] ++
+      [ (1, genIFix) | k == Star ] ++
+      [ (5, genLam k1 k2) | KindArrow _ k1 k2 <- [k] ] ++
+      [ (5, genApp) ]
+  where
+
+    checkInvariants gen = do
+      ty <- gen
+      debug <- asks geDebug
+      ctx <- asks geTypes
+      if debug then
+        case checkKind ctx ty k of
+          Left err ->
+             (error . show $ "genType - checkInvariants: type " <> prettyPirReadable ty
+                           <> " does not match kind " <> prettyPirReadable k
+                           <> " in context " <> prettyPirReadable ctx
+                           <> " with error message " <> fromString err)
+          _ -> return ty
+      else
+        return ty
+
+    -- this size split keeps us from generating riddiculous types that
+    -- grow huge to the left of an arrow or abstraction (See also the
+    -- genApp case below). This ratio of 1:7 was not scientifically
+    -- established, if you are unhappy about the compleixty of the
+    -- type of arguments that are generated tweaking this might
+    -- be a good idea.
+    genFun = uncurry (TyFun ()) <$> sizeSplit_ 1 7 (genType k) (genType k)
+
+    genForall = do
+      x <- genMaybeFreshTyName "a"
+      k <- liftGen arbitrary
+      fmap (TyForall () x k) $ onSize (subtract 1) $ bindTyName x k $ genType Star
```

Comment:
What do you mean by constant here?

Reply:
This is not what this code does - `onSize` here explicitly does *not* touch the `QuickCheck` size parameter but our own `geSize` (as per the note).

Reply:
Oh, I think I got it backwards! I thought `geSize` was for controlling the size of constants rather than the size of the AST and QuickCheck's size parameter was used for the latter, but it's exactly the opposite. Is that correct? If so, I'll update the Note a little for easily confusable people like me.

Reply:
Oh, I think I got it backwards! I thought `geSize` was for controlling the size of constants rather than the size of the AST and QuickCheck's size parameter was used for the latter, but it's exactly the opposite. Is that correct? If so, I'll update the Note a little for easily confusable people like me.

Reply:
Thats right, the note could probably be clearer (that is, I think the note is at fault not you).